### PR TITLE
Fix #196: Crypto 3.0: Update sharedInfo1 for ECIES

### LINKED ISSUE #196: Crypto 3.0: Update sharedInfo1 for ECIES
ECIES scheme for Crypto 3.0 takes (optional) `sharedInfo1` and concatenates it with `ephemeralPublicKey`. The `BasicEciesEncryptor` and `BasicEciesDecryptor` classes should include this logic.

I suggest to always apply this concatentation logic before deriving the secret key using KDF for Crypto 3.0.

We can still support old `sharedInfo1` (we need it for `createToken` SOAP method in version 2.0) by using empty `sharedInfo1`, which will still lead to just the `ephemeralPublicKey` being used as `sharedInfo1` for KDF.

This way we do not have to introduce new methods for `v3` and we can still support `v2`.

### PR DESCRIPTION
**Pull request by romanstrobl**
Changes:
- Updated `sharedInfo1` parameter based on Crypto 3.0 ECIES scheme
- Move ECIES envelope key derivation and storage into class EciesEnvelopeKey
- Unified class names, method names and parameters with mobile SDK
- Implemented factory for encryptors and decryptors with correct parameters for application and activation scopes

### CODE REVIEW

File: powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesDecryptor.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+     *     <li>2.1</li>
+     * </ul>
+     *
+     * For PowerAuth version 3.0 use {@link #encryptResponse(byte[])} because request data and MAC is always be available.
+     *
+     * @param data Response data to encrypt.
+     * @return ECIES cryptogram.
+     * @throws EciesException In case response encryption fails.
+     */
+    public EciesCryptogram encryptResponseDirect(byte[] data, byte[] ephemeralPublicKeyBytes) throws EciesException {
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesDecryptor.java]: Yes. One day, but not yet.

File: powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesDecryptor.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Decrypt request data from cryptogram.
+     *
+     * @param cryptogram ECIES cryptogram.
+     * @return Decrypted data.
+     * @throws EciesException In case request decryption fails.
+     */
+    public byte[] decryptRequest(EciesCryptogram cryptogram) throws EciesException {
+        this.envelopeKey = EciesEnvelopeKey.fromPrivateKey(privateKey, cryptogram.getEphemeralPublicKey(), sharedInfo1);
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesDecryptor.java]: Yes, thanks for catching this one.

File: powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesEncryptor.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    /**
+     * Encrypt request data.
+     *
+     * @param data Request data.
+     * @return ECIES cryptogram.
+     * @throws EciesException In case request encryption fails.
+     */
+    public EciesCryptogram encryptRequest(byte[] data) throws EciesException {
+        this.envelopeKey = EciesEnvelopeKey.fromPublicKey(publicKey, sharedInfo1);
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/EciesEncryptor.java]: Yes, thanks for catching this one.

File: powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/model/SharedInfo1Constants.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Enumeration with constants for parameter sharedInfo1.
+ */
+public enum SharedInfo1Constants {
+
+    ACTIVATION_LAYER_2("/pa/activation"),
+    MIGRATION_START("/pa/migration"),
+    VAULT_UNLOCK("/pa/vault/unlock"),
+    CREATE_TOKEN("/pa/token/create");
```
[the remark this answers is not in the document]
Comment (hvge) [reply powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/model/SharedInfo1Constants.java]: I've added two new constants for generic ECIES setup for app & activation scope: `/pa/generic/application`, `/pa/generic/activation`. Check out our private TLDR doc with ongoing changes for details.

**Pull request merged**